File: 2024/R/solve_day3_part1.R
#' Sum of multiplications
#'
#' Uses regex to find numbers in format mul(x, y). Multiplies each pair and
#' adds the results together.
#'
#' @param input
#' Character vector of lines to parse for numbers to multiply.
#'
#' @return
#' The sum of results of multiplications
solve_day3_part1 <- function(input) {
  muls <- stringi::stri_extract_all_regex(input,
                                          "(?<=mul\\()[0-9]+,[0-9]+(?=\\))")

  muls <- unlist(muls)
  muls <- strsplit(muls, ",")
  muls <- purrr::list_transpose(muls)
  muls <- lapply(muls, as.numeric)

  sum(muls[[1]] * muls[[2]])
}

File: 2024/python/solve_day3_part1.py
import re

def solve_day3_part1(text):
  """
  Sum of multiplications
  
  Uses regex to find numbers in format mul(x, y). Multiplies each pair and
  adds the results together.
  
  Parameters
  ----------
  text : list of str
    Lines to parse for numbers to multiply.
      
  Returns
  -------
  int
    The sum of results of multiplications
  """
  muls = [re.findall("(?<=mul\()[0-9]+,[0-9]+(?=\))", line) for line in text]
  muls = [pair for line in muls for pair in line] 
  muls = [pair.split(",") for pair in muls]
  muls = [[int(num) for num in pair] for pair in muls]
  muls = [pair[0] * pair[1] for pair in muls]
  
  return(sum(muls))
